lms/services/document_url.py

```diff
@@ -0,0 +1,107 @@
+import re
+from functools import lru_cache
+from typing import Optional
+from urllib.parse import unquote
+
+from lms.services.vitalsource import VitalSourceService
+
+
+class DocumentURLService:
+    """A service for getting LTI launch document URLs."""
+
+    def __init__(self, assignment_service):
+        self._assignment_service = assignment_service
+
+    @lru_cache(1)
+    def get_document_url(self, context, request) -> Optional[str]:
+        """
+        Get the configured document for this LTI launch.
+
+        This will try all known methods for finding out, and will return None
+        if no configuration can be found.
+        """
+        for document_url_source in (
+            self._from_deep_linking_provided_url,
+            self._from_canvas_file,
+            self._from_legacy_vitalsource_book,
+            self._from_assignment_in_db,
+        ):
+            if document_url := document_url_source(context, request):
+                return document_url
+
+        return None
+
+    _ENCODED_URL = re.compile("^(?:https?|canvas|vitalsource)%3a", re.IGNORECASE)
+
+    @classmethod
+    def _from_deep_linking_provided_url(cls, _context, request):
+        """Get the URL from our own configuration set during deep linking."""
```

Comment:
`our own configuration in the launch URL`?

Reply:
I was trying to say that we set this value, it's not something the LMS decides to do on it's own.

Maybe this would read better as 

```suggestion
        """Get the URL from the configuration we send to the LMS during deep linking."""
```

... or something that fits on the line etc.